program-runtime/src/loaded_programs.rs

```diff
@@ -0,0 +1,551 @@
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+impl solana_frozen_abi::abi_example::AbiExample for LoadedPrograms {
+    fn example() -> Self {
+        // Delegate AbiExample impl to Default before going deep and stuck with
+        // not easily impl-able Arc<dyn Executor> due to rust's coherence issue
+        // This is safe because LoadedPrograms isn't serializable by definition.
+        Self::default()
+    }
+}
+
+impl LoadedPrograms {
+    /// Inserts a single entry
+    pub fn insert_entry(&mut self, key: Pubkey, entry: LoadedProgram) -> bool {
+        let second_level = self.entries.entry(key).or_insert_with(Vec::new);
+        let index = second_level
+            .iter()
+            .position(|at| at.effective_slot >= entry.effective_slot);
+        if let Some(index) = index {
+            let existing = second_level
+                .get(index)
+                .expect("Missing entry, even though position was found");
+            if existing.deployment_slot == entry.deployment_slot
+                && existing.effective_slot == entry.effective_slot
+            {
+                return false;
+            }
+        }
+        second_level.insert(index.unwrap_or(second_level.len()), Arc::new(entry));
+        true
+    }
+
+    /// Before rerooting the blockstore this removes all programs of orphan forks
+    pub fn prune<F: ForkGraph>(&mut self, fork_graph: &F, new_root: Slot) {
+        self.entries.retain(|_key, second_level| {
+            let mut first_ancestor = true;
+            *second_level = second_level
+                .iter()
+                .rev()
+                .filter(|entry| {
+                    let relation = fork_graph.relationship(entry.deployment_slot, new_root);
+                    if entry.deployment_slot >= new_root {
+                        matches!(relation, BlockRelation::Equal | BlockRelation::Descendant)
+                    } else if first_ancestor {
+                        first_ancestor = false;
+                        matches!(relation, BlockRelation::Ancestor)
+                    } else {
+                        false
+                    }
+                })
+                .cloned()
+                .collect();
+            second_level.reverse();
+            !second_level.is_empty()
+        });
+    }
+
+    /// Extracts a subset of the programs relevant to a transaction batch
+    /// and returns which program accounts the accounts DB needs to load.
+    pub fn extract<S: WorkingSlot>(
+        &self,
+        working_slot: &S,
+        keys: impl Iterator<Item = Pubkey>,
+    ) -> (HashMap<Pubkey, Arc<LoadedProgram>>, Vec<Pubkey>) {
+        let mut missing = Vec::new();
+        let found = keys
+            .filter_map(|key| {
+                if let Some(second_level) = self.entries.get(&key) {
+                    for entry in second_level.iter().rev() {
+                        if working_slot.is_ancestor(entry.deployment_slot) {
```

Comment:
Also needs to check if `entry.effective_slot` is visible yet.

Reply:
Updated the code, and added some tests.